OpenSimRT/RealTime/include/RealTimeAnalysis.h

```diff
@@ -0,0 +1,224 @@
+        std::vector<InverseKinematics::IMUTask> ikIMUTasks;
+        double ikConstraintsWeight;
+        double ikAccuracy;
+
+        // id + jr parameters
+        std::vector<ExternalWrench::Parameters> wrenchParameters;
+
+        // so parameters
+        bool solveMuscleOptimization;
+        MuscleOptimization::OptimizationParameters muscleOptimizationParameters;
+        MomentArmFunctionT momentArmFunction;
+    };
+
+    struct Loggers {
+        // ik
+        OpenSim::TimeSeriesTable qLogger;
+        OpenSim::TimeSeriesTable qDotLogger;
+        OpenSim::TimeSeriesTable qDDotLogger;
+
+        // id
+        OpenSim::TimeSeriesTable tauLogger;
+
+        // so
+        OpenSim::TimeSeriesTable fmLogger;
+        OpenSim::TimeSeriesTable amLogger;
+        OpenSim::TimeSeriesTable residualLogger;
+
+        // jr
+        OpenSim::TimeSeriesTable jrLogger;
+    };
+
+ public:
+    // ctor
+    RealTimeAnalysis(const OpenSim::Model& model, const Parameters& parameters);
+    virtual ~RealTimeAnalysis() = default; // dtor
+
+    /**
+     * Start the simulation. It creates one thread for acquiring data, solving
+     * IK and filtering, and one thread for performing the rest of the analysis
+     * (ID, SO and JR). The threads are detached and terminated when simulation
+     * ends or signaled from the main thread.
+     */
+    void run();
+
+    /**
+     * Check the termination flag if it has been raised.
+     */
+    bool shouldTerminate();
+
+    /**
+     * Set the termination flag to terminate analysis threads.
+     */
+    void shouldTerminate(bool flag);
+
+    /**
+     * Thread safe fetch function of analysis results.
+     */
+    Output getResults();
+
+    /**
+     * Initialize module loggers.
+     */
+    Loggers initializeLoggers();
+
+ protected:
+    /**
+     * This function is meant to be used in a separate thread to handle the data
+     * acquisition.
+     */
+    virtual void acquisition();
+
+    /**
+     * This function is meant to be used in a separate thread to handle the data
+     * processing.
+     */
+    virtual void processing();
+
+    /**
+     * Prepare the input data for filtering.
+     */
+    SimTK::Vector prepareUnfilteredData(
+            const SimTK::Vector& q,
+            const std::vector<ExternalWrench::Input>& externalWrenches) const;
+
+    OpenSim::Model model;
+    Parameters parameters; // RealTimeAnalysis parameters
+    Loggers log;           // loggers
+    Output output;         // analysis results
+    double previousAcquisitionTime;
+    double previousProcessingTime;
+
+    // modules
+    SimTK::ReferencePtr<LowPassSmoothFilter> lowPassFilter;
+    SimTK::ReferencePtr<InverseKinematics> inverseKinematics;
+    SimTK::ReferencePtr<InverseDynamics> inverseDynamics;
+    SimTK::ReferencePtr<MuscleOptimization> muscleOptimization;
+    SimTK::ReferencePtr<JointReaction> jointReaction;
+
+    // data buffer
+    CircularBuffer<1, LowPassSmoothFilter::Output> buffer;
```

Comment:
Maybe we have a larger buffer in case we change the behavior in the future, but process only the last frame in the processing thread. What do you think?

Reply:
Holds resources for no reason. I don't see the point in this. I could leave a note, or create a #define with the buffer size.